### PR TITLE
feat: BasePaymentService migration

### PR DESCRIPTION
**Pull request by adrien2p**
Move the BasePaymentService to the main medusa package and migrate the service to TS

**Goal**
- Move the `BasePaymentService` to the core
- Re work the naming of the method to be more understandable
- Move the service to an abstract class

**Information**
One that PR will be merged, then we will be able to go forward on the refactoring of the core as well as the payment plugins

### CODE REVIEW

**Comment by adrien2p** [comment]
@srindom 
Should the methods be called:
- `createPayment`-> `getPaymentSessionData`: This method is actually called by `createSession` and expect `paymentSessionData` the name should be explicit such as `getPaymentData` which is call on `createPayment`
- `updatePayment` -> `getUpdatedPaymentSessionData`:  This method is actually called by `updateSession` and expect `paymentSessionData`

File: packages/medusa/src/payments/payment-service.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import { Payment } from "@models/payment"
+import { Cart } from "@models/cart"
+import { PaymentSession, PaymentSessionStatus } from "@models/payment-session"
+import { Customer } from "@models/customer"
+import { BaseService } from "medusa-interfaces"
+
+export type Data = Record<string, unknown>
+export type PaymentData = Data
+export type PaymentSessionData = Data
+
```
Comment (adrien2p) [inline packages/medusa/src/payments/payment-service.ts:10]: The three types above are the same but increase readability about there purpose. It will be easier for a user to understand the intelisens when seeing `PaymentSessionData` instead of just `Record`

File: packages/medusa/src/payments/payment-service.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import { Payment } from "@models/payment"
```
Comment (srindom) [inline packages/medusa/src/payments/payment-service.ts:1]: Can we rename this file to `src/interfaces/payment-service.ts`

File: packages/medusa/src/payments/payment-service.ts (hunk 1 above)
Comment (srindom) [reply packages/medusa/src/payments/payment-service.ts:10]: I think the types here are an excellent way of doing things! 🎉 

In regards to your other comment about naming of the methods: I think we need to stick with `create...` when we are initializing a payment session so that we are indicating to plugin authors that "this is the place you create something; i.e. it is okay to make a post request here".

If we call it `getPaymentSessionData` it feels like it is only okay to do retrieval of existing data. What are your thoughts on maybe naming it this way:

`createPayment` -> `createPaymentSessionData` or maybe even `onCreatePaymentSession` to indicate that it is more like a handle?

Comment (olivermrbl) [reply packages/medusa/src/payments/payment-service.ts:10]: > I think the types here are an excellent way of doing things! 🎉

I agree! Would it make sense to add `export type Data = Record<string, unknown>` to our shared, since I could imagine, that we would be using this typing pattern elsewhere.

**Comment by adrien2p** [comment]
I proposed `onCreatePaymentSession` at the beginning so its fine for me :)

File: packages/medusa/src/interfaces/payment-service.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import { BaseService } from "medusa-interfaces"
+
+export type Data = Record<string, unknown>
+export type PaymentData = Data
+export type PaymentSessionData = Data
+
+export interface PaymentService {
+  getIdentifier(): string
+
+  onCreatePaymentSession(car: Cart): Promise<PaymentSessionData>
```
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-service.ts:14]: nit:
```suggestion
  onCreatePaymentSession(cart: Cart): Promise<PaymentSessionData>
```

File: packages/medusa/src/interfaces/payment-service.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  protected static identifier: string
+
+  public getIdentifier(): string {
+    if (!(<typeof AbstractPaymentService>this.constructor).identifier) {
+      throw new Error('Missing static property "identifier".')
+    }
+    return (<typeof AbstractPaymentService>this.constructor).identifier
+  }
+
+  public abstract onCreatePaymentSession(car: Cart): Promise<PaymentSessionData>
```
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-service.ts:54]: nit:
```suggestion
  public abstract onCreatePaymentSession(cart: Cart): Promise<PaymentSessionData>
```

File: packages/medusa/src/interfaces/payment-service.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  public abstract onCreatePayment(
+    paymentSession: PaymentSession
+  ): Promise<PaymentData>
+
+  public abstract retrievePayment(car: Cart): Promise<Payment>
```
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-service.ts:60]: nit: 
```suggestion
  public abstract retrievePayment(cart: Cart): Promise<Payment>
```

File: packages/medusa/src/interfaces/payment-service.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+export type PaymentSessionData = Data
+
+export interface PaymentService {
+  getIdentifier(): string
+
+  onCreatePaymentSession(car: Cart): Promise<PaymentSessionData>
+
+  onCreatePayment(paymentSession: PaymentSession): Promise<PaymentData>
+
+  retrievePayment(car: Cart): Promise<Payment>
```
Comment (olivermrbl) [inline packages/medusa/src/interfaces/payment-service.ts:18]: nit: 
```suggestion
  retrievePayment(cart: Cart): Promise<Payment>
```

**Comment by adrien2p** [comment]
@srindom anything else from your point of view ?

File: packages/medusa/src/interfaces/payment-service.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  onCreatePayment(paymentSession: PaymentSession): Promise<PaymentData>
+
+  retrievePayment(cart: Cart): Promise<Payment>
+
+  updatePayment(
+    paymentSessionData: PaymentSessionData,
+    cart: Cart
+  ): Promise<PaymentSessionData>
+
+  authorizePayment(
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:25]: Should we name this one `onAuthorizePayment`?

File: packages/medusa/src/interfaces/payment-service.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    paymentSession: PaymentSession,
+    context: Data
+  ): Promise<{ data: PaymentSessionData; status: PaymentSessionStatus }>
+
+  capturePayment(payment: Payment): Promise<PaymentData>
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:30]: `onCapturePayment`

File: packages/medusa/src/interfaces/payment-service.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+  refundPayment(payment: Payment, refundAmount: number): Promise<PaymentData>
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:32]: `onRefundPayment`

File: packages/medusa/src/interfaces/payment-service.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+  deletePayment(paymentSession: PaymentSession): Promise<void>
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:34]: `onDeletePayment`

File: packages/medusa/src/interfaces/payment-service.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+export interface PaymentService {
+  getIdentifier(): string
+
+  onCreatePaymentSession(cart: Cart): Promise<PaymentSessionData>
+
+  onCreatePayment(paymentSession: PaymentSession): Promise<PaymentData>
+
+  retrievePayment(cart: Cart): Promise<Payment>
+
+  updatePayment(
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:20]: `onUpdatePayment`

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:20]: This one would be more `onUpdateSession`

File: packages/medusa/src/interfaces/payment-service.ts (hunk 10 above)
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:34]: this one is called with `refreshSession` so probably keeping `deletePayment` to keep the action as it is ?

File: packages/medusa/src/interfaces/payment-service.ts (hunk 9 above)
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:32]: it really is the action to refund the payment with the external provider, it is not just building dome data on refund action so I think keeping the name is fine

File: packages/medusa/src/interfaces/payment-service.ts (hunk 8 above)
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:30]: it really is the action to capture the payment with the external provider, it is not just building some data on capture action so I think keeping the name is fine

File: packages/medusa/src/interfaces/payment-service.ts (hunk 7 above)
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:25]: it really is the action to authorizing the payment with the external provider, it is not just building some data on authorized action so I think keeping the name is fine

**Comment by adrien2p** [comment]
Unlike the methods we renamed prefixed with `on...` where those methods where doing complementary actions and there new name keeps the meaning, I dont think it is necessarily true with the other methods. For example, `OnRefreshSession` does not insinuate that we must delete the payment? What do you think?

**Comment by adrien2p** [comment]
```typescript
getIdentifier(): string

  initiatePaymentSession(cart: Cart): Promise<PaymentSessionData>

  retrievePaymentData(paymentSession: PaymentSession): Promise<PaymentData>

  updatePaymentSession(
    paymentSessionData: PaymentSessionData,
    cart: Cart
  ): Promise<PaymentSessionData>

  authorizePayment(
    paymentSession: PaymentSession,
    context: Data
  ): Promise<{ data: PaymentSessionData; status: PaymentSessionStatus }>

  capturePayment(payment: Payment): Promise<PaymentData>

  refundPayment(payment: Payment, refundAmount: number): Promise<PaymentData>

  deletePayment(paymentSession: PaymentSession): Promise<void>

  retrieveSavedMethods(customer: Customer): Promise<Data[]>

  getStatus(): PaymentSessionStatus
  ```
  
  What do you think like that ?

File: packages/medusa/src/interfaces/payment-service.ts (hunk 7 above)
Comment (srindom) [reply packages/medusa/src/interfaces/payment-service.ts:25]: By this standard, however, I feel that `onCreatePaymentSession` is also being named incorrectly as `onCreatePaymentSession` in most cases will be an actual call to a 3rd party API to "initialize" a session, e.g. Stripe creates a PaymentIntent, Klarna creates a Checkout instance etc. 

Maybe let's circle back once again and look at what we are trying to do here:
- Choose names that clearly indicate to a plugin author what the logic in the function should do
- Strengthen the type safety and versatility of the interface

To the first point, I think both `onAuthorizePayment` and `authorizePayment` are equally good at explaining what happens - the `on`-prefix could to a certain extent indicate more clearly that the method is being called as part of some other logic orchestration (namely the Cart service, when it is trying transforming or verifying that a payment is authorized) although I don't think that in and of itself makes it really clear. Another thing that I think is quite important in the interface is that we are consistent so if we name one method `onXYZ` then I think we should name the rest `onABC`.

With all this being said maybe we should go back to dropping `on` from all function names and just make sure that they are correctly explaining what they are creating or what they are doing: e.g. `createPaymentSession`, `authorizePaymentSession`, `updatePaymentSession`, `createPayment`, `capturePayment`, etc.

**Comment by adrien2p** [comment]
@srindom I let you have a look and if its ok for you, we can start migrating the plugins

**Comment by adrien2p** [comment]
@srindom do we agree on that PR ?

**Comment by adrien2p** [comment]
@srindom once that PR is merged, we should then create an issue to
- Update the payment-provider service
- Update all payment provider that extender the paymentService
- Update the doc

File: packages/medusa/src/interfaces/payment-service.ts [hunk 12]
```diff
@@ -0,0 +1,107 @@
+import { Payment } from "@models/payment"
+import { Cart } from "@models/cart"
+import { PaymentSession, PaymentSessionStatus } from "@models/payment-session"
+import { Customer } from "@models/customer"
+import { BaseService } from "medusa-interfaces"
```
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:5]: todo: use the new BaseService

**Comment by adrien2p** [comment]
@srindom what do you think of that approach for the decorators? https://github.com/adrien2p/medusa-extender/tree/main/src/decorators
And here is the entry point to load them and them per type https://github.com/adrien2p/medusa-extender/blob/main/src/loaders/modules.loader.ts

Let me know what is your though about that approach

**Comment by adrien2p** [comment]
@srindom I will start to integrate the migration then

**Comment by adrien2p** [comment]
@srindom I think that you will be to start the review again. Please, have a strong look at the payment plugins in order to avoid regression, I know you know them properly. 😸 
If everything is fine, then we can re think of the method naming or let them like this at the moment and open another pr later on

File: packages/medusa/src/loaders/plugins.ts [hunk 13]
```diff
@@ -252,14 +251,15 @@ export async function registerServices(pluginDetails: PluginDetails, container:
...
+      const isService = loaded.prototype instanceof BaseService || loaded.prototype instanceof TransactionBaseService
+      if (!isService) {
         const logger = container.resolve<Logger>("logger")
         const message = `Services must inherit from BaseService, please check ${fn}`
         logger.error(message)
         throw new Error(message)
       }
 
-      if (loaded.prototype instanceof PaymentService) {
+      if (loaded.prototype instanceof AbstractPaymentService) {
```
Comment (srindom) [inline packages/medusa/src/loaders/plugins.ts:262]: thought: wouldn't it be best to make it possible for "legacy" payment providers to still work?

Comment (srindom) [reply packages/medusa/src/loaders/plugins.ts:262]: And then remove the old interface in a coming release?

File: packages/medusa/src/models/payment.ts [hunk 14]
```diff
@@ -72,10 +72,10 @@ export class Payment {
   data: any
 
   @Column({ type: resolveDbType("timestamptz"), nullable: true })
-  captured_at: Date
+  captured_at: Date | string
```
Comment (srindom) [inline packages/medusa/src/models/payment.ts:75]: question: why is this necessary?

**Review by srindom — COMMENTED** [review]
Looks great - will try out all the payment plugins; in the meantime would be good to hear your thoughts on a way in which we could make the old payment service work also :)

File: packages/medusa/src/loaders/plugins.ts (hunk 13 above)
Comment (adrien2p) [reply packages/medusa/src/loaders/plugins.ts:262]: Yes I can definitely do that, and depending on which one is used the payment provider service can use the appropriate method.

And i will decorate it as deprecated then

File: packages/medusa/src/models/payment.ts (hunk 14 above)
Comment (adrien2p) [reply packages/medusa/src/models/payment.ts:75]: Because a date can also be a string. For example toISOString will be a string 😅 generally speaking a date is always a date is a string and then when you use it you always pass it to the date constructor for later manipulation

**Comment by adrien2p** [comment]
Generally speaking, this is where the pattern adapter can be handy. It allows you to have a common api from different api. But since this is fairly rare and will be removed in later version, i think we will go for an easier solution that check weather it is one or the other

**Comment by adrien2p** [comment]
Ok the changes has been done, the api is the same for now for both base service. The naming will come later when we will agree on that.

File: packages/medusa-payment-adyen/src/services/adyen.js [hunk 15]
```diff
@@ -1,12 +1,12 @@
 import axios from "axios"
 import _ from "lodash"
 import { hmacValidator } from "@adyen/api-library"
-import { BaseService } from "medusa-interfaces"
+import { TransactionBaseService, PaymentSessionStatus } from "@medusajs/medusa"
 import { Client, Config, CheckoutAPI } from "@adyen/api-library"
 
-class AdyenService extends BaseService {
+class AdyenService extends TransactionBaseService {
```
Comment (olivermrbl) [inline packages/medusa-payment-adyen/src/services/adyen.js:7]: issue: Should extend the abstract payment service
```suggestion
class AdyenService extends AbstractPaymentService {
```

File: packages/medusa-payment-adyen/src/services/adyen.js [hunk 16]
```diff
@@ -150,12 +135,12 @@ class AdyenService extends BaseService {
 
   /**
    * Status for Adyen payment.
-   * @param {object} paymentData - payment method data from cart
-   * @returns {string} the status of the payment
+   * @param {Record<string, unknown>} paymentData - payment method data from cart
+   * @returns {Promise<PaymentSessionStatus>} the status of the payment
    */
-  getStatus(paymentData) {
+  async getStatus(paymentData) {
```
Comment (olivermrbl) [inline packages/medusa-payment-adyen/src/services/adyen.js:141]: question: This method does not need to be an async one as far as I can see. What's your reasoning behind adding it?

File: packages/medusa-payment-klarna/src/services/klarna-provider.js [hunk 17]
```diff
@@ -246,29 +246,28 @@ class KlarnaProviderService extends PaymentService {
...
   /**
    * Status for Klarna order.
-   * @param {Object} paymentData - payment method data from cart
-   * @returns {string} the status of the Klarna order
+   * @param {PaymentSessionData} paymentSessionsData - payment session data from cart
+   * @returns {Promise<PaymentSessionStatus>} the status of the Klarna order
    */
-  async getStatus(paymentData) {
-    const { order_id } = paymentData
+  async getStatus(paymentSessionsData) {
```
Comment (olivermrbl) [inline packages/medusa-payment-klarna/src/services/klarna-provider.js:252]: very, very nit(non-blocking): Should be in singular form
```suggestion
  async getStatus(paymentSessionData) {
```

File: packages/medusa-payment-stripe/src/services/stripe-bancontact.js [hunk 18]
```diff
@@ -1,15 +1,13 @@
-import _ from "lodash"
 import Stripe from "stripe"
-import { PaymentService } from "medusa-interfaces"
+import { AbstractoPaymentService } from "@medusajs/medusa"
+import { PaymentSessionData } from "@medusajs/medusa/src/interfaces"
+import { Cart } from "@medusajs/medusa/src/models"
 
-class BancontactProviderService extends PaymentService {
+class BancontactProviderService extends AbstractoPaymentService {
```
Comment (olivermrbl) [inline packages/medusa-payment-stripe/src/services/stripe-bancontact.js:6]: issue: Typo in import. Is your IDE not complaining about it? 🤔 
```suggestion
class BancontactProviderService extends AbstractPaymentService {
```

File: packages/medusa/src/services/payment-provider.ts [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+  async retrievePayment(
+    id: string,
+    relations: string[] = []
+  ): Promise<Payment> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const paymentRepo = transactionManager.getCustomRepository(
+          this.paymentRepository_
+        )
+        const validatedId = validateId(id)
```
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:93]: todo(non-blocking): Remove since config is not passed i.e. simply returning the provided id.

File: packages/medusa/src/services/payment-provider.ts [hunk 20]
```diff
@@ -0,0 +1,597 @@
...
+  async retrieveSession(
+    id: string,
+    relations: string[] = []
+  ): Promise<PaymentSession> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const sessionRepo = transactionManager.getCustomRepository(
+          this.paymentSessionRepository_
+        )
+        const validatedId = validateId(id)
```
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:146]: See other comment.

File: packages/medusa/tsconfig.json [hunk 21]
```diff
@@ -16,7 +16,13 @@
...
     "allowJs": true,
     "skipLibCheck": true,
-    "downlevelIteration": true // to use ES5 specific tooling
+    "downlevelIteration": true, // to use ES5 specific tooling
+    "baseUrl": ".",
+    "paths": {
+      "@models/*": [
+        "./src/models/*"
+      ]
+    }
```
Comment (olivermrbl) [inline packages/medusa/tsconfig.json:25]: question: Is this a needed?

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Really great job @adrien2p! Have left a couple of comments and requested some changes 👍

File: packages/medusa-payment-stripe/src/services/stripe-bancontact.js (hunk 18 above)
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/services/stripe-bancontact.js:6]: No 😂 it is a js file maybe thats why it does not say anything, the pipeline did not get any trouble to build as well

File: packages/medusa/tsconfig.json (hunk 21 above)
Comment (adrien2p) [reply packages/medusa/tsconfig.json:25]: No I will remove it and it could be part of another pr. It is to avoid infinite ../../../

File: packages/medusa-payment-adyen/src/services/adyen.js (hunk 16 above)
Comment (adrien2p) [reply packages/medusa-payment-adyen/src/services/adyen.js:141]: Basically, the getSatus from the payment-provider was already async, so I thought that giving the ability to the consumer to have some third party call on async processing of the status could be something interesting. But I might be wrong, let me know wdyt

Comment (olivermrbl) [reply packages/medusa-payment-adyen/src/services/adyen.js:141]: Got it! Also fine with keeping it async 👍

**Review by olivermrbl — APPROVED** [review]
LGTM 🚀 

Just need @srindom to have a look as well!

**Comment by srindom** [comment]
Any chance that we can split this PR into 1 pr per provider?
